test/utils/bundle-browser.js

```diff
@@ -78,6 +78,10 @@ class Node extends libp2p {
 
     defaultsDeep(_options, defaults)
 
+    // TODO: is there a defaultsDeep that doesn't copy the refs?
+    _options.modules.peerDiscovery.push(wrtcStar.discovery)
+    _options.modules.peerDiscovery.push(wsStar.discovery)
```

Comment:
Is there a defaultsDeep that doesn't clone objects that it doesn't need to clone/merge?

Reply:
@alanshaw, @pgte, @satazor, @jacobheun et al, calling out for your JS wisdom to help on this one :)

Reply:
You can use `https://lodash.com/docs/4.17.10#mergeWith` with a customizer function that does the "defaults" logic by checking for undefineds.

Reply:
I think mergeWith will still clone the objects. You could use something like [deepmerge](https://github.com/KyleAMathews/deepmerge) and turn cloning off. The array merging would need to get updated for it though. For example, the following would combine options and defaults. Any arrays that exist in both, the options array would win out. This should be all object references, no clones.

```js
_options = deepmerge({_options, defaults, { clone: false, arrayMerge: (destinationArray, sourceArray, options) => sourceArray })
```

Reply:
@jacobheun I just tried it and it doesn't:

<img width="450" alt="screen shot 2018-06-19 at 11 36 59" src="https://user-images.githubusercontent.com/1017236/41592651-2908197a-73b5-11e8-88b3-b35656a653a8.png">

The lodash documentation says the following:

> Array and plain object properties are merged recursively

Reply:
Ah yep, you're right. Also just did a test with mergeWith for custom merging and it looks solid.

Reply:
Ideally, we want to replace the defaultsDeep + the later array.push by something simpler. Currently, I don't see that using another package with multiple options and custom merge functions is better.

Reply:
I agree that abstracting that out to a simple method call is the way to go, so we can use that elsewhere and have consistent defaults behavior, if there are no existing modules that do it the way we need it to behave. I think the issue with doing array.push later is that if users want to specify their own discovery array and not use those two defaults, they will have to remove them after creating the libp2p node. 

As a user, I would expect that if I pass options for `modules.peerDiscovery`, I wouldn't get the defaults. If I want to add additional peerDiscovery options, I'd expect to have to add them after node creation. If those rules are intended to be different I think they need to be explicitly called out.